### PR TITLE
KTrajectoryPulseq

### LINKED ISSUE #72: Pulseq Trajectory
As I haven't found an issue for this:

We plan to have a  seq-File+ismrmrd->KData path.

This would include using the .seq to calculator the trajectory (main task)
And maybe filling in some header information into misc.

@Stef-Martin  is currently looking into it.

### PR DESCRIPTION
**Pull request by JoHa0811**
Calculate trajectory from pulseq sequence file

Closes #72

### CODE REVIEW

File: src/mrpro/data/traj_calculators/_KTrajectorySeq.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        k0 = int(k_traj_adc.shape[1] / k1 / k2)
+
+        k_traj_adc[0] = k_traj_adc[0] / np.max(np.abs(k_traj_adc[0])) * np.pi
+        k_traj_adc[1] = k_traj_adc[1] / np.max(np.abs(k_traj_adc[1])) * np.pi
+        k_traj_adc[2] = k_traj_adc[2] / np.max(np.abs(k_traj_adc[2])) * np.pi
+
+        kx = torch.tensor(k_traj_adc[0]).view((1, k2, k1, k0))
+        ky = torch.tensor(k_traj_adc[1]).view((1, k2, k1, k0))
+        kz = torch.tensor(k_traj_adc[2]).view((1, k2, k1, k0))
+        kx = rearrange(kx, 'other k2 k1 k0 -> (other k2 k1) k0', k2=k2, k1=k1, other=1)
```
Comment (fzimmermann89) [inline src/mrpro/data/traj_calculators/_KTrajectorySeq.py:72]: other should not be 1, but k0 shoud be k0

File: src/mrpro/data/traj_calculators/_KTrajectorySeq.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        # k0 : num k space points per spiral
+
+        seq = pp.Sequence()
+        seq.read(file_path=self.path)
+        k_traj_adc, _, _, _, _ = seq.calculate_kspace()
+
+        unique_idxs = {label: np.unique(getattr(kheader.acq_info.idx, label)) for label in ['k1', 'k2']}
+        k1 = len(unique_idxs['k1'])
+        k2 = len(unique_idxs['k2'])
+        k0 = int(k_traj_adc.shape[1] / k1 / k2)
```
Comment (fzimmermann89) [inline src/mrpro/data/traj_calculators/_KTrajectorySeq.py]: could also be wrong if there are any other counters set

Comment (fzimmermann89) [reply src/mrpro/data/traj_calculators/_KTrajectorySeq.py:63]: still a bit confused what shape this should return in the end...

**Comment by ckolbPTB** [comment]
Supersedes #78

File: src/mrpro/data/traj_calculators/_KTrajectoryPulseq.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        """
+
+        # create PyPulseq Sequence object and read .seq file
+        seq = pp.Sequence()
+        seq.read(file_path=str(self.seq_path))
+
+        # calculate k-space trajectory using PyPulseq
+        k_traj_adc, _, _, _, _ = seq.calculate_kspace()
+
+        unique_idxs = {label: np.unique(getattr(kheader.acq_info.idx, label)) for label in ['k1', 'k2']}
```
Comment (schuenke) [inline src/mrpro/data/traj_calculators/_KTrajectoryPulseq.py]: before the merge all numpy functions should be replaced by torch functions.

File: src/mrpro/data/traj_calculators/_KTrajectoryPulseq.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+        # get number of samples as integer
+        # ToDo: find more pythonic solution compatible with mypy
+        k0 = int(num_samples[0].squeeze().tolist()[0])
+
+        sample_size = num_samples.shape[0]
+
+        k_traj_adc[0] = k_traj_adc[0] / np.max(np.abs(k_traj_adc[0])) * np.pi
+        k_traj_adc[1] = k_traj_adc[1] / np.max(np.abs(k_traj_adc[1])) * np.pi
+        k_traj_adc[2] = k_traj_adc[2] / np.max(np.abs(k_traj_adc[2])) * np.pi
```
Comment (schuenke) [inline src/mrpro/data/traj_calculators/_KTrajectoryPulseq.py]: replace numpy -> torch

**Review by schuenke — CHANGES_REQUESTED** [review]
replace all numpy functions with torch functions, than we can merge it.

File: src/mrpro/data/_KData.py [hunk 5]
```diff
@@ -149,7 +148,7 @@ def reshape_acq_data(data):
             case KTrajectoryCalculator():
                 ktraj_calc = ktrajectory(kheader)
                 if isinstance(ktraj_calc, KTrajectoryRawShape):
-                    ktraj = ktraj_calc.reshape(sort_idx, num_k2, num_k1)
+                    ktraj = ktraj_calc.reshape(sort_idx, num_k2, num_k1, repeat_detection_tolerance=None)
```
Comment (ckolbPTB) [inline src/mrpro/data/_KData.py]: Why is `repeat_detection_tolerance=None` needed? @schuenke I think this caused the problems you had when trying out Andreas' `FourierOp`

File: src/mrpro/data/_KData.py [hunk 6]
```diff
@@ -162,7 +161,7 @@ def reshape_acq_data(data):
 
         try:
             shape = ktraj.broadcasted_shape
-            torch.broadcast_shapes(kdata.shape, shape)
+            torch.broadcast_shapes(kdata[:, 0].shape, shape)
```
Comment (ckolbPTB) [inline src/mrpro/data/_KData.py]: Why is this needed?

File: src/mrpro/data/_KHeader.py [hunk 7]
```diff
@@ -1,3 +1,4 @@
+# %%
```
Comment (ckolbPTB) [inline src/mrpro/data/_KHeader.py]: Needs to be removed

File: src/mrpro/data/_KTrajectoryRawShape.py [hunk 8]
```diff
@@ -64,5 +64,4 @@ def reshape(
         kz = rearrange(self.kz[sort_idx, ...], '(other k2 k1) k0 -> other k2 k1 k0', k1=num_k1, k2=num_k2)
         ky = rearrange(self.ky[sort_idx, ...], '(other k2 k1) k0 -> other k2 k1 k0', k1=num_k1, k2=num_k2)
         kx = rearrange(self.kx[sort_idx, ...], '(other k2 k1) k0 -> other k2 k1 k0', k1=num_k1, k2=num_k2)
-
```
Comment (ckolbPTB) [inline src/mrpro/data/_KTrajectoryRawShape.py]: Please undo this change

File: src/mrpro/data/_KData.py (hunk 6 above)
Comment (JoHa0811) [reply src/mrpro/data/_KData.py]: Sanity check if trajectory and kdata match

File: src/mrpro/data/_KData.py (hunk 5 above)
Comment (JoHa0811) [reply src/mrpro/data/_KData.py]: in _KTrajectory there is a check for this parameter. We had to set it to None. Default was a float. I honestly cannot tell you anymore why exactly it has to be done..

Comment (JoHa0811) [reply src/mrpro/data/_KData.py]: actually seems to work without as well. removed it now

File: src/mrpro/data/_KData.py (hunk 6 above)
Comment (JoHa0811) [reply src/mrpro/data/_KData.py]: actually seems to work without as well. removed it now

**Review by schuenke — APPROVED** [review]
